builtin/credential/aws/path_login.go

```diff
@@ -337,11 +355,144 @@ func (b *backend) parseIdentityDocument(s logical.Storage, pkcs7B64 string) (*id
 	return &identityDoc, nil
 }
 
-// pathLoginUpdate is used to create a Vault token by the EC2 instances
+func (b *backend) pathLoginUpdate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	anyEc2, allEc2 := hasValuesForEc2Auth(data)
+	anyIam, allIam := hasValuesForIamAuth(data)
+	switch {
+	case anyEc2 && anyIam:
+		return logical.ErrorResponse("supplied auth values for both ec2 and iam auth types"), nil
+	case anyEc2 && !allEc2:
+		return logical.ErrorResponse("supplied some of the auth values for the ec2 auth type but not all"), nil
+	case anyEc2:
+		return b.pathLoginUpdateEc2(req, data)
+	case anyIam && !allIam:
+		return logical.ErrorResponse("supplied some of the auth values for the iam auth type but not all"), nil
+	case anyIam:
+		return b.pathLoginUpdateIam(req, data)
+	default:
+		return logical.ErrorResponse("didn't supply required authentication values"), nil
+	}
+}
+
+// Returns whether the EC2 instance meets the requirements of the particular
+// AWS role entry.
+func (b *backend) verifyInstanceMeetsRoleRequirements(
+	s logical.Storage, instance *ec2.Instance, roleEntry *awsRoleEntry, roleName string, iamClient *iam.IAM) (*roleTagLoginResponse, string, error) {
+
```

Comment:
With that, I suppose the `iamClient` creation bits can also move into this function.